drivers/sensor/bq274xx/bq274xx.c

```diff
@@ -636,26 +637,156 @@ static int bq274xx_gauge_configure(const struct device *dev)
+{
+	const struct bq274xx_config *const config = dev->config;
+	int status = 0;
+
+	if (!device_is_ready(config->gpio_pm.port)) {
+		LOG_DBG("GPIO device pointer is not ready to be used");
+		return -EINVAL;
+	}
+
+	status = gpio_pin_configure_dt(&config->gpio_pm,
+			   GPIO_OUTPUT | GPIO_OPEN_DRAIN);
+	if (status < 0) {
+		LOG_ERR("Unable to configure interrupt pin to output and open drain");
+		return status;
+	}
+
+	status = gpio_pin_set(config->gpio_pm.port, config->gpio_pm.pin, 0);
+	if (status < 0) {
+		LOG_ERR("Unable to set interrupt pin to low");
+		return status;
+	}
+
+	k_msleep(1);
+
+	status = gpio_pin_configure_dt(&config->gpio_pm,
+			   GPIO_INPUT);
+	if (status < 0) {
+		LOG_ERR("Unable to configure interrupt pin to input");
+		return status;
+	}
+
+	k_msleep(100);
+
+	status = bq274xx_gauge_configure(dev);
+	if (status < 0) {
+		LOG_ERR("Unable to configure bq274xx gauge");
+		return status;
+	}
+
 	return 0;
 }
 
+static int bq274xx_pm_control(const struct device *dev, uint32_t ctrl_command,
+				  void *context, device_pm_cb cb, void *arg)
+{
+	int ret = 0;
+	struct bq274xx_data *data = (struct bq274xx_data *)dev->data;
+
+	switch (ctrl_command) {
+	case DEVICE_PM_SET_POWER_STATE:
+		if (*((uint32_t *)context) == DEVICE_PM_OFF_STATE) {
+			ret = bq274xx_enter_shutdown_mode(data);
+			if (ret < 0) {
+				LOG_ERR("Unable to enter off state");
+			}
+		} else {
+			ret = bq274xx_exit_shutdown_mode(dev);
+			if (ret < 0) {
+				LOG_ERR("Unable to enter active state");
+			}
+		}
+		break;
+	case DEVICE_PM_GET_POWER_STATE:
+		*((uint32_t *)context) = data->pm_state;
+		break;
+	default:
+		ret = -EINVAL;
+	}
+
+	if (cb != NULL) {
+		cb(dev, ret, context, arg);
+	}
+	return ret;
+}
+#endif /* CONFIG_PM_DEVICE */
+
 static const struct sensor_driver_api bq274xx_battery_driver_api = {
 	.sample_fetch = bq274xx_sample_fetch,
 	.channel_get = bq274xx_channel_get,
 };
 
+#ifdef CONFIG_PM_DEVICE
+#define BQ274XX_INT_CFG(index)						      \
+	.gpio_pm = GPIO_DT_SPEC_INST_GET(index, int_gpios),
+#else
+#define BQ274XX_INT_CFG(index)
+#endif
+
 #define BQ274XX_INIT(index)                                                    \
 	static struct bq274xx_data bq274xx_driver_##index;                     \
 									       \
 	static const struct bq274xx_config bq274xx_config_##index = {          \
+		BQ274XX_INT_CFG(index)                                         \
 		.bus_name = DT_INST_BUS_LABEL(index),                          \
 		.design_voltage = DT_INST_PROP(index, design_voltage),         \
 		.design_capacity = DT_INST_PROP(index, design_capacity),       \
 		.taper_current = DT_INST_PROP(index, taper_current),           \
 		.terminate_voltage = DT_INST_PROP(index, terminate_voltage),   \
 	};                                                                     \
 									       \
-	DEVICE_DT_INST_DEFINE(index, &bq274xx_gauge_init, NULL,                \
+	DEVICE_DT_INST_DEFINE(index, &bq274xx_gauge_init, bq274xx_pm_control,  \
```

Comment:
What happens if you disable CONFIG_PM_DEVICE? I guess this one will generate build problems, isn't it? You need to keep it under ifdef as well.

Reply:
Funnily enough, when I remove the `CONFIG_PM_DEVICE` from my `prj.conf` file, but leave the `bq274xx_pm_control` nothing breaks. Also, when I was first writing the `bq274xx_pm_control` function I was looking through examples like `drivers/sensor/qdec_nrfx` and `drivers/sensor/lis2mdl` and they all had structure the same as mine.

So yeah, I don't know why it isn't throwing any errors.

Reply:
Because in DEVICE_DT_INST_DEFINE expansion (see include/device.h) if CONFIG_PM_DEVICE is not defined then the pm_init routine is not used. See below: 

```
#ifdef CONFIG_PM_DEVICE                                                         
#define Z_DEVICE_DEFINE_PM_INIT(dev_name, pm_control_fn)>       >       \           
>       .pm_control = (pm_control_fn),> >       >       >       \               
>       .pm = &Z_DEVICE_STATE_NAME(dev_name).pm,                                
#else                                                                           
#define Z_DEVICE_DEFINE_PM_INIT(dev_name, pm_control_fn)                            
#endif                                                                          
```

So, the code is fine. Good!